cornac/models/causalrec/recom_causalrec.py

```diff
@@ -0,0 +1,400 @@
+                
+                if self.tanh < 2:
+                    l2_e = _l2_loss(E, Bp, E_ind)
+                else:
+                    l2_e = _l2_loss(E, Bp, E_ind, E2, E_ind2)
+                
+                reg = (
+                        _l2_loss(gamma_u, gamma_i, gamma_j, theta_u) * self.lambda_w
+                        + _l2_loss(beta_i) * self.lambda_b
+                        + _l2_loss(beta_j) * self.lambda_b / 10
+                        + l2_e * self.lambda_e
+                )
+                
+                loss = -log_likelihood + reg - log_likelihood_m - log_likelihood_n
+                
+                optimizer.zero_grad()
+                loss.backward()
+                optimizer.step()
+                
+                sum_loss += loss.data.item()
+                count += len(batch_u)
+                if count % (self.batch_size * 10) == 0:
+                    progress_bar.set_postfix(loss=(sum_loss / count))
+                progress_bar.update(1)
+            progress_bar.close()
+        
+        print("Optimization finished!")
+        
+        self.beta_item = Bi.data.cpu().numpy()
+        self.gamma_user = Gu.data.cpu().numpy()
+        self.gamma_item = Gi.data.cpu().numpy()
+        self.theta_user = Tu.data.cpu().numpy()
+        self.emb_matrix = E.data.cpu().numpy()
+        self.beta_prime = Bp.data.cpu().numpy()
+        self.emb_dir_matrix = E_ind.data.cpu().numpy()
+        # pre-computed for faster evaluation
+        self.theta_item = F.mm(E)
+        if self.tanh == 0:
+            self.theta_item = self.theta_item.data.cpu().numpy()
+        elif self.tanh == 1:
+            self.theta_item = torch.tanh(self.theta_item).data.cpu().numpy()
+        elif self.tanh == 2:
+            self.theta_item = torch.tanh(torch.tanh(self.theta_item).mm(E2)).data.cpu().numpy()
+
+        self.visual_bias = F.mm(Bp).squeeze().data.cpu().numpy()
+
+        self.ind_theta_item = F.mm(E_ind)
+        if self.tanh == 0:
+            self.ind_theta_item = self.ind_theta_item.data.cpu().numpy()
+        elif self.tanh == 1:
+            self.ind_theta_item = torch.tanh(self.ind_theta_item).data.cpu().numpy()
+        elif self.tanh == 2:
+            self.ind_theta_item = torch.tanh(torch.tanh(self.ind_theta_item).mm(E_ind2)).data.cpu().numpy()
+
+        self.beta_item_mean = Bi.mean().unsqueeze(dim=0).data.cpu().numpy()
+        self.gamma_item_mean = Gi.mean(dim=0).unsqueeze(dim=0).data.cpu().numpy()
+
+        self.mean_feat = mean_feat.mm(E_ind)
+        if self.tanh == 0:
+            self.mean_feat = self.mean_feat.data.cpu().numpy()
+        elif self.tanh == 1:
+            self.mean_feat = torch.tanh(self.mean_feat).data.cpu().numpy()
+        elif self.tanh == 2:
+            self.mean_feat = torch.tanh(torch.tanh(self.mean_feat).mm(E_ind2)).data.cpu().numpy()
+
+    def score(self, user_idx, item_idx=None, vb=0.8):
```

Comment:
Is it be better to move the `vb` param to the model constructor and be part of the model hyper-params? This will help standardize the score function, also less confusing for the users.

Reply:
Definitely. In the last commit, I have changed the naming of this variable to ```lambda_2``` and move it to the constructor with a documentation.